components/feature/downloads/src/test/java/mozilla/components/feature/downloads/DownloadsFeatureTest.kt

```diff
@@ -6,321 +6,343 @@ package mozilla.components.feature.downloads
-                fragmentManager = mockFragmentManager
-            )
+    fun `Calling stop() will unregister listeners from download manager`() {
+        val downloadManager: DownloadManager = mock()
+
+        val feature = DownloadsFeature(
+            testContext,
+            store,
+            useCases = mock(),
+            downloadManager = downloadManager
+        )
 
         feature.start()
-        assertTrue(feature.dialog !is SimpleDownloadDialogFragment)
-        verify(mockSession).download
+
+        verify(downloadManager, never()).unregisterListeners()
+
+        feature.stop()
+
+        verify(downloadManager).unregisterListeners()
     }
 
     @Test
-    fun `shouldn't show twice a dialog if is already created`() {
-        val mockDialog = spy(DownloadDialogFragment::class.java)
-        val mockFragmentManager = mock<FragmentManager>()
+    fun `DownloadManager failing to start download will cause error toast to be displayed`() {
+        grantPermissions()
 
-        `when`(mockFragmentManager.findFragmentByTag(FRAGMENT_TAG)).thenReturn(mockDialog)
+        val downloadManager: DownloadManager = mock()
+        doReturn(
+            arrayOf(INTERNET, WRITE_EXTERNAL_STORAGE)
+        ).`when`(downloadManager).permissions
 
-        val featureWithDialog =
-            DownloadsFeature(
-                testContext,
-                downloadManager = mockDownloadManager,
-                sessionManager = mockSessionManager,
-                fragmentManager = mockFragmentManager
-            )
+        doReturn(null).`when`(downloadManager).download(any(), anyString())
 
-        grantPermissions()
+        val feature = spy(DownloadsFeature(
+            testContext,
+            store,
+            useCases = mock(),
+            downloadManager = downloadManager
+        ))
+
+        doNothing().`when`(feature).showDownloadNotSupportedError()
+
+        feature.start()
+
+        verify(downloadManager, never()).download(any(), anyString())
+        verify(feature, never()).showDownloadNotSupportedError()
 
-        featureWithDialog.start()
+        val download: DownloadState = mock()
+        store.dispatch(ContentAction.UpdateDownloadAction("test-tab", download))
+            .joinBlocking()
 
-        startDownload()
+        testDispatcher.advanceUntilIdle()
 
-        verify(mockDialog, times(0)).show(mockFragmentManager, FRAGMENT_TAG)
+        verify(downloadManager).download(eq(download), anyString())
+        verify(feature).showDownloadNotSupportedError()
     }
 
     @Test
-    fun `download is cleared when permissions denied`() {
-        feature.start()
-        feature.onPermissionsResult(arrayOf(INTERNET, WRITE_EXTERNAL_STORAGE),
-                intArrayOf(PermissionChecker.PERMISSION_GRANTED, PermissionChecker.PERMISSION_GRANTED))
-        assertNull(mockSessionManager.selectedSession)
+    fun `showDownloadNotSupportedError shows toast`() {
+        Dispatchers.resetMain()
```

Comment:
Yeah. I should add a comment here, I think. The reason for that is that showing a `Toast` will crash the test since we are not actually calling it from the main thread (with a `Looper` set up etc.) with the `TestDispatcher` setup. I reset here to actually run on the real main thread and revert what we did in `setUp()`.